### ISSUE TITLE
Add support for "session"-duration secret cookies

### ISSUE THREAD

**Issue by ecton** [SergioBenitez/Rocket#1506]
I'm currently using the `master` branch to develop a new project against, and in the process of setting up the Remember Me checkbox for sign in, I noticed I wasn't able to get the cookie to be stored with a 'Session' duration. This is [done by not emitting an Expires attribute](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie). While that page also notes a warning about session restore causing issues, I still for my site want my cookies to be able to be tied to a Session so that Remember Me emits a session-length cookie when unchecked and a long expiration time when Remember Me is checked.

I discovered that this isn't currently possible. `CookieJar::add_private` calls `CookieJar::set_private_defaults`, which has this chunk:

```rust
if cookie.expires().is_none() {
  cookie.set_expires(time::OffsetDateTime::now_utc() + time::Duration::weeks(1));
}
```

This prevents inserting a private cookie with no expiration date. I actually completely agree that setting a duration by default is a good security tactic, but in my case, I want to explicitly ensure this cookie has no Expires attribute.

For `SameSite` this is done by offering a `SameSite::None` in addition to allowing `None` when setting the same_site attribute. I would suggest offering some sort of similar value that can represent an explicitly omitted Expires attribute.

**Comment by SergioBenitez** [comment]
I think we should keep these defaults but also provide a kind of escape hatch. One idea is to add a couple more methods to the effect of `add_session_cookie` and `add_private_session_cookie`. Another is to have a specific setting in `Cookie` itself to make it a "session" cookie, abstracting away _what_ makes it a session cookie. Here, we'd hope something like the following would have your desired effect:

```rust
let cookie = Cookie::build("foo", "bar").session();
cookies.add_private(cookie);
```

There is precedent for this given the existence of the `permanent` builder method.

@jebrosen What do you think?

**Comment by jebrosen** [comment]
It seems like one of the underlying problems here is that `expires` is an `Option<OffsetDateTime>`, so there's no way to distinguish between "not set explicitly" and "set to None / session duration". This seems fine for `Cookie` in isolation; it can say that the default is `None` / session duration, but for setting private cookie defaults we would ideally not override an explicitly-set `None`. Most of the other values where `rocket_http` overrides the defaults *do* have a way to distinguish between "unset" and "set-to-`None`".

> Another is to have a specific setting in Cookie itself to make it a "session" cookie, abstracting away what makes it a session 

This sounds reasonable on the face of it, but I worry that it would lead to two APIs ("session()" and "expires()") that inspect/mutate the same source of truth. I think I could be okay with something like that, as long as the interactions between those two and potential gotchas are made very clear.

**Comment by arnodb** [comment]
Hi Rocketers,

That is something I faced with at work and here is the current state of what we did:

- With version 0.4 we just reimplemented the function which adds private cookies and removed the default value, such a fork is acceptable as its size is very limited.
- With version 0.5 the API does now allow such a fork and the problem remains "to fix". So I'll keep looking at that particular issue updates. Whatever you do to allow session private cookies, I'll take it :).

Side note: our migration to 0.5 is on standby due to many things moving on your side but all in all I did not spot anything that would stop us doing that.
